### PR TITLE
[New Scheduler]Implement PFCInvokerServer

### PR DESCRIPTION
**Pull request by ningyougang**
## Description
Manage New scheduler's invokerServer. provide basic disable/enable for the invoker.

These http endpoints are for deployment or get the invoker's current situation, nomally, when do zerodowntime deployment,
need to send /disable request to invoker, make the invoker's status to `down`, then, the new activations will not be sent to the invoker, we can wait the invoker executed all flying activations, then, we can stop the invoker safely and deploy the invoker .

* /disable
  e.g. curl -X POST -u xxx:xxx http://$invoker:12001/disable, disabled the invoker, the invoker status will be changed to `unhealthy` firstly and changed to`down` finally after several seconds.
* /enable
  e.g. curl -X POST -u xxx:xxx http://$invoker:12001/enable, enable the invoker, the invoker status will be changed to `up` again.

* /memory (this is for FPCInvokerServer only)
  e.g. curl -X GET -u xxx:xxx http://$invoker:12001/memory, if the result >0, the invoker is busy, otherwise, it is idle.

Due to simple, no document.

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Scheduler
- [x] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

File: core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/Invoker.scala [hunk 1]
```diff
@@ -227,9 +231,3 @@ trait InvokerServerProvider extends Spi {
   def instance(
     invoker: InvokerCore)(implicit ec: ExecutionContext, actorSystem: ActorSystem, logger: Logging): BasicRasService
 }
-
-object DefaultInvokerServer extends InvokerServerProvider {
```
Comment (ningyougang) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/Invoker.scala:231]: I moved DefaultInvokerServer to a separate file， keep the same as FPCInvokerServer.

File: core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/InvokerReactive.scala [hunk 2]
```diff
@@ -293,10 +295,52 @@ class InvokerReactive(
...
+  override def disable(): Route = {
+    scheduler match {
+      case Some(schedulerActor) =>
+        // Through negative number of invoker to pass unhealthy message
+        healthProducer
+          .send(
+            "health",
+            PingMessage(
+              InvokerInstanceId(
+                -1 - instance.instance,
```
Comment (ningyougang) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/InvokerReactive.scala:328]: Use the same topic name `health` to send unhealthy message via netative number of invoker.
In controller side(e.g. InvokerSupervision), get the message from kafka and make the invoker to `unhealthy` and `down` finally.

File: core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/InvokerSupervision.scala [hunk 3]
```diff
@@ -175,8 +175,26 @@ class InvokerPool(childFactory: (ActorRefFactory, InvokerInstanceId) => ActorRef
     val raw = new String(bytes, StandardCharsets.UTF_8)
     PingMessage.parse(raw) match {
       case Success(p: PingMessage) =>
-        self ! p
-        invokerPingFeed ! MessageFeed.Processed
+        if (p.instance.instance < 0) {
```
Comment (ningyougang) [inline core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/InvokerSupervision.scala:178]: if user wants to do some zerodowntime deployment, need to disable the invoker firstly and wait invoker executed all flying activations.

User can send /disable to the invoker,  in invoker backend, use the same topic name `health` to send unhealthy message via negative number of invoker.

here, get the message from kafka and make the invoker to `unhealthy` and `down` finally.

File: core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/FPCInvokerServer.scala [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  implicit val jsonPrettyResponsePrinter = PrettyPrinter
+
+  override def routes(implicit transid: TransactionId): Route = {
+    super.routes ~ extractCredentials {
+      case Some(BasicHttpCredentials(username, password)) if username == systemUsername && password == systemPassword =>
+        (path("enable") & post) {
+          invoker.enable()
+        } ~ (path("disable") & post) {
+          invoker.disable()
+        } ~ (path("memory") & get) {
```
Comment (ningyougang) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/FPCInvokerServer.scala:34]: Regarding `/memory`, this is for FPCInvokerServer only, this is used for `judge whether existing flying activations on the invoker`, if the result(busyMemory + inProgressMemory) > 0, it shows the invoker is executing activations, if the result == 0, it shows the invoker is idle.

**Comment by ningyougang** [comment]
Have any comment ?

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthManager.scala [hunk 5]
```diff
@@ -269,6 +270,7 @@ object InvokerHealthManager {
   val healthActionNamePrefix = "invokerHealthTestAction"
   val bufferSize = 10
   val bufferErrorTolerance = 3
+  var useMemory = 0l
```
Comment (style95) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthManager.scala:273]: Why do we need this mutable variable?

Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthManager.scala:273]: There has a get used memory API: http://$invoker:$port/memory, 
```
complete(InvokerHealthManager.useMemory.toString)
```
The FunctionPullingContainerPool actor(which will be contributed) sends the pool's used memory info(busyMemory + inprogressMemory) to InvokerHealthyManager actor, so use a mutable variable to store the value.

File: core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/DefaultInvokerServer.scala [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+          invoker.disable()
+        }
+      case _ => terminate(StatusCodes.Unauthorized)
+    }
+  }
+}
+
+object DefaultInvokerServer extends InvokerServerProvider {
+
+  // TODO: TBD, after FPCInvokerReactive is ready, can read the credentials from pureconfig
```
Comment (bdoyle0182) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/DefaultInvokerServer.scala:58]: Not a problem for this mr, but I don't think that's a secure way of setting admin credentials

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthManager.scala (hunk 5 above)
Comment (bdoyle0182) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthManager.scala:273]: nit: `usedMemory`

Comment (bdoyle0182) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthManager.scala:273]: I know that's there only one instance of the invoker health manager per instance / jvm, but it still seems risky to be mutating a var in an object like this especially when it's the source of truth on whether the invoker is idle or not.

**Comment by bdoyle0182** [comment]
LGTM other than the discussion on the mutable variable

File: core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/DefaultInvokerServer.scala (hunk 6 above)
Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/DefaultInvokerServer.scala:58]: Yes, after FPCInvokerReactive is ready, can read the credentials from pureconfig like other components.

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthManager.scala (hunk 5 above)
Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthManager.scala]: @style95 @bdoyle0182 , yes, not a good idea of `use mutable variable`
It is better to do like below
```
functionPullingContainerPool.ask(UsedMemory)(Timeout(5.seconds)).mapTo[Int]
```
But functionPullingContainerPool is implementing now, so i removed /memory api in this pr and i will add /memory api in ` [New Scheduler] FunctionPullingContainerPool` pr.

Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthManager.scala]: Ok, good catch, as i said above, i removed /memory api in this pr. i will add `/memory api` in another pr: `[New Scheduler] FunctionPullingContainerPool` with `functionPullingContainerPool.ask(UsedMemory)(Timeout(5.seconds)).mapTo[Int]`.

**Comment by ningyougang** [comment]
Have any more comment?

File: core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/InvokerReactive.scala [hunk 7]
```diff
@@ -299,4 +300,8 @@ class InvokerReactive(
     }
   })
 
+  override def enable(): Route = ???
```
Comment (style95) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/InvokerReactive.scala:303]: I haven't tried it but would it be ok to leave these methods empty to run it?

Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/InvokerReactive.scala:303]: Sorry, you mean implement it like below?
```
override def enable(): Route = {
    complete("todo or pong")
}
```
Due to InvokerReactive extends InvokerCore which has `enabled/disabled`,
so need to implement enable/disabled

**Review by bdoyle0182 — APPROVED** [review]

**Review by style95 — APPROVED** [review]

File: core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/InvokerReactive.scala (hunk 7 above)
Comment (jiangpengcheng) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/InvokerReactive.scala:303]: when deploy using `InvokerReactive` and send `enable/disable` requests to invoker, what will happen?

**Comment by ningyougang** [comment]
One of the unit test case is failed all the time.
image

I pull above images in my local successfully, don't know the concrete reason why in travis is failed.

**Review by jiangpengcheng — APPROVED** [review]

**Pull request merged**